fem/libceed/algebraic.cpp

```diff
@@ -0,0 +1,627 @@
+// Copyright (c) 2010-2020, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#include "algebraic.hpp"
+
+#ifdef MFEM_USE_CEED
+#include "../bilinearform.hpp"
+#include "../fespace.hpp"
+#include "../libceed/solvers-atpmg.h"
+#include "../libceed/full-assembly.hpp"
+#include "../pfespace.hpp"
+
+namespace mfem
+{
+
+Solver *BuildSmootherFromCeed(ConstrainedMFEMCeedOperator &op, bool chebyshev)
+{
+   int ierr;
+   CeedOperator ceed_op = op.GetCeedOperator();
+   const Array<int> &ess_tdofs = op.GetEssentialTrueDofs();
+   const Operator *P = op.GetProlongation();
+   // Assemble the a local diagonal, in the sense of L-vector
+   CeedVector diagceed;
+   CeedInt length;
+   ierr = CeedOperatorGetSize(ceed_op, &length); PCeedChk(ierr);
+   ierr = CeedVectorCreate(internal::ceed, length, &diagceed); PCeedChk(ierr);
+   CeedMemType mem;
+   ierr = CeedGetPreferredMemType(internal::ceed, &mem); PCeedChk(ierr);
+   if (!Device::Allows(Backend::CUDA) || mem != CEED_MEM_DEVICE)
+   {
+      mem = CEED_MEM_HOST;
+   }
+   Vector local_diag(length);
+   CeedScalar *ptr = (mem == CEED_MEM_HOST) ? local_diag.HostWrite() :
+                     local_diag.Write(true);
+   ierr = CeedVectorSetArray(diagceed, mem, CEED_USE_POINTER, ptr);
+   PCeedChk(ierr);
+   ierr = CeedOperatorLinearAssembleDiagonal(ceed_op, diagceed,
+                                             CEED_REQUEST_IMMEDIATE);
+   PCeedChk(ierr);
+   ierr = CeedVectorTakeArray(diagceed, mem, NULL); PCeedChk(ierr);
+
+   Vector t_diag;
+   if (P)
+   {
+      t_diag.SetSize(P->Width());
+      P->MultTranspose(local_diag, t_diag);
+   }
+   else
+   {
+      t_diag.NewMemoryAndSize(local_diag.GetMemory(), length, false);
+   }
+   Solver *out = NULL;
+   if (chebyshev)
+   {
+      const int cheb_order = 3;
+      out = new OperatorChebyshevSmoother(&op, t_diag, ess_tdofs, cheb_order);
+   }
+   else
+   {
+      const double jacobi_scale = 0.65;
+      out = new OperatorJacobiSmoother(t_diag, ess_tdofs, jacobi_scale);
+   }
+   ierr = CeedVectorDestroy(&diagceed); PCeedChk(ierr);
+   return out;
+}
+
+#ifdef MFEM_USE_MPI
+
+class CeedAMG : public Solver
+{
+public:
+   CeedAMG(ConstrainedMFEMCeedOperator &oper, HypreParMatrix *P)
```

Comment:
I certainly don't need it, it is here just for modularity/clarity. Note that it is only in the `.cpp` file and is not a user-facing class.

The idea is to separate ownership and creation/destruction of some objects so that IMO `AlgebraicCeedMultigrid` is simpler and easier to understand. Would you like it better if it was a `BuildAMGFromCeed()` function analogous to the `BuildSmootherFromCeed()` function above?

Reply:
I understand the scope and purpose now.